### PR TITLE
Add Synchronous APIs to ObjectSerializer and JsonSerializer

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR adds synchronous APIs to the `ObjectSerializer` and `JsonSerializer` interfaces. In addition APIs that return `OutputStream` were changed to return an extension of `OutputStream` based on the type of `OutputStream` passed into the API.

Asynchronous APIs will wrap the synchronous APIs as serialization and deserialization code works in a synchronous manner. This change was done as adding synchronous on top of asynchronous in the future could lead to Reactor throwing exceptions for blocking in a reactive thread.

### CODE REVIEW

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/JsonSerializer.java [hunk 1]
```diff
@@ -12,6 +12,17 @@
...
+    /**
+     * Reads a JSON stream into its object representation.
+     *
+     * @param stream JSON stream.
+     * @param typeReference {@link TypeReference} representing the object.
+     * @param <T> Type of the object.
+     * @return The object represented by the deserialized JSON stream.
+     */
+    @Override
+    <T> T deserializeSync(InputStream stream, TypeReference<T> typeReference);
```
Comment (srnagar) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/JsonSerializer.java:24]: Probably should inverse the naming for sync and async the way we do for clients. `deserialize` and `deserializerAsync`.

Comment (alzimmermsft) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/JsonSerializer.java:24]: My thought process here was we want the asynchronous APIs to be the main scenario, so I decided to give the synchronous APIs the worse name.

@JonathanGiles, thoughts as well?

Comment (JonathanGiles) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/JsonSerializer.java:24]: Yeah, I would make it Async - it looks a little odd with the Sync suffix.

Comment (alzimmermsft) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/serializer/JsonSerializer.java]: Updated naming to use `Async` suffix for asynchronous methods, synchronous methods have no suffix.

**Pull request merged**